lib/deepcrawl.js

```diff
@@ -0,0 +1,137 @@
+'use strict';
+const BPromise = require('bluebird'),
+  schema = require('./schema'),
+  needle = BPromise.promisifyAll(require('needle')),
+  humps = require('humps'),
+  _ = require('lodash');
+
+class DeepCrawl {
+  constructor(cfg) {
+    cfg = cfg || {};
+    if (!cfg.apiId) {
+      throw new Error('DeepCrawl called without required arg "apiId"');
+    }
+    if (!cfg.apiKey) {
+      throw new Error('DeepCrawl called without required arg "apiKey"');
+    }    
+
+    this.apiId = cfg.apiId;
+    this.apiKey = cfg.apiKey;
+    this.getSessionToken = cfg.getSessionToken;
+    this.accountId = cfg.accountId;
+    this.baseUrl = cfg.baseUrl || 'https://prod-1-dc-api-oopeix3r.deepcrawl.com';
+  }
+
+  createEndpoints (schema) {
+    const API = {};
+    for (const endpoint in schema) {
+      API[endpoint] = {};
+    }
+
+    return API;
+  }
+
+  generateRequest (url, options, method) {
+    // first, get the sessionToken and generate the request options and headers
+    return BPromise.resolve(options.sessionToken || this.getSessionToken())
+      .then((sessionToken) => {
+        const requestOpts = {
+          json: true,
+          headers: {
+            'X-Auth-Token': sessionToken
+          }
+        };
+
+        if (method === 'get') {
+          return needle.getAsync(url, requestOpts)
+            .then(this.handleResponse);
+        }
+
+        // convert all camelCased options to _ for deepcrawl
+        const data = humps.decamelizeKeys(options);
+        return needle[`${method}Async`](url, data, requestOpts)
+          .then(this.handleResponse);
+      });
+  }
+
+  generateMethod (requiredFields, url, requestMethod) {
+    return (options) => {
+      options = options || {};
+      if (!options.sessionToken && !this.getSessionToken) {
+        throw new Error(`"sessionToken" or "getSessionToken" method required`);
+      }
+
+      requiredFields.forEach((field) => {
+        if (!options[field] && !this[field]) {
+          throw new Error(`${field} is required`);
+        }
+        // if the url contains {field}, replace {field} with the passed
+        // in value
+        if (url.indexOf(`{${field}}`) > -1) {
+          url = url.replace(`{${field}}`, (options[field] || this[field]));
+        }
+      });
+
+      return this.generateRequest(url, options, requestMethod);     
+    };
+  }
+
+  addMethodsToEndpoints (API, schema) {
+    for (const endpoint in API) {
+      const endpointUrl = this.baseUrl + schema[endpoint].route,
+        requiredFields = _.clone(schema[endpoint].requiredFields || []),
+        id = schema[endpoint].id;
+      // remove the endpoint route, required fields, and id from the schema
+      delete schema[endpoint].requiredFields;
+      delete schema[endpoint].route;
+      delete schema[endpoint].id;
```

Comment:
Shouldn't need to delete these. Per my comments below you should be able to just do `for (const action in schema[endpoint].actions)`